text/0000-constants-in-patterns.md

```diff
@@ -0,0 +1,316 @@
+[motivation]: #motivation
+
+The main motivation to write this RFC is to finish what started in [RFC 1445][rfc-1445]: define what happens when a constant is used as a pattern.
+That RFC is incomplete in several ways:
+
+- It was never fully implemented; due to bugs in the early implementation, parts of it are still behind future-compatibility lints.
+  This also leads to a rather messy situation in rustc where const-in-pattern handling has to deal with fallback cases and emitting four (!) such lints.
+  We should clean up both our language specification and the implementation in rustc.
+- The RFC said that matching on floats should be fully rejected, but when a PR was made to enforce this, [many people spoke up against that and it got rejected](https://github.com/rust-lang/rust/pull/84045).
+- The RFC does not explain how to treat raw pointers and function pointers.
+
+[rfc-1445]: https://rust-lang.github.io/rfcs/1445-restrict-constants-in-patterns.html
+
+RFC 1445 had the goal of leaving it open whether we want constants in patterns to be treated like sugar for primitive patterns or for `PartialEq`-based equality tests.
+This new RFC takes the stance it does the former based on the following main design goals:
+
+- Refactoring a pattern that has no binders, wildcards, or ranges into a constant should never change behavior.
+  This aligns with the oft-repeated intuition that a constant works "as if" its value was just copy-pasted everywhere the constant is used.
+  This is particularly important for patterns that syntactically look exactly like constants, namely zero-field enum variants. Consider:
+
+    ```rust
+    enum E { Var1, Var2 }
+
+    const BEST_VAR: E = E::Var1;
+
+    fn is_best(e: E) -> bool { matches!(e, BEST_VAR) }
+    fn is_var1(e: E) -> bool { matches!(e, E::Var1) }
+    ```
+
+    It would be very surprising if those two functions could behave differently.
+  It follows that we cannot allow `PartialEq` to affect the behavior of constants in patterns.
+
+- We do not want to expose equality tests on types where the library author did not explicitly expose an equality test.
+  This means not allowing matching on constants whose type does not implement `PartialEq`.
+  It also means not allowing matching on constants where running `PartialEq` would behave different from the corresponding pattern:
+  the pattern can be accessing private fields, and the `PartialEq` implementation provided by the library might be treating those fields in a particular way;
+  we should not let people write patterns that "bypass" any such treatment and do structural matching when the crate author does not provide a structural `PartialEq`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Constants can be used as patterns, but only if their type implements `PartialEq`.
+Moreover, this implementation must be the automatically derived one, and that also applies recursively for the types of their fields (recursively):
```

Comment:
Is one of these "recursively"s redundant?

I'd expect "applies recursively to their fields" to mean fields of fields (recursively), and "applies to their (top-level) fields" to mean just the fields named in the struct.

```suggestion
Moreover, this implementation must be the automatically derived one, and that also applies recursively for the types of their fields:
```

Reply:
Oh yeah, that's redundant.^^